token-swap/program/src/curve/constant_product.rs

```diff
@@ -43,6 +44,93 @@ pub fn swap(
     })
 }
 
+/// Get the amount of trading tokens for the given amount of pool tokens,
+/// provided the total trading tokens and supply of pool tokens.
+///
+/// The constant product implementation is a simple ratio calculation for how many
+/// trading tokens correspond to a certain number of pool tokens
+pub fn pool_tokens_to_trading_tokens(
+    pool_tokens: u128,
+    pool_token_supply: u128,
+    swap_token_a_amount: u128,
+    swap_token_b_amount: u128,
+    round_direction: RoundDirection,
+) -> Option<TradingTokenResult> {
+    let mut token_a_amount = pool_tokens
+        .checked_mul(swap_token_a_amount)?
+        .checked_div(pool_token_supply)?;
+    let mut token_b_amount = pool_tokens
+        .checked_mul(swap_token_b_amount)?
+        .checked_div(pool_token_supply)?;
+    let (token_a_amount, token_b_amount) = match round_direction {
+        RoundDirection::Floor => (token_a_amount, token_b_amount),
+        RoundDirection::Ceiling => {
+            let token_a_remainder = pool_tokens
+                .checked_mul(swap_token_a_amount)?
+                .checked_rem(pool_token_supply)?;
+            if token_a_remainder > 0 && token_a_amount > 0 {
```

Comment:
Why check for `token_a_amount > 0` here? Isn't it valid for `token_a_amount` to be ~0.99 and round up?

Reply:
Since we're doing a ceiling and not a rounding, we can run into bad UX if someone asks for 1 pool token, which is worth 0.01 token A, and take 1 token A from them for it.  There will always be some lost value in some direction, but this check avoids the bad offenders at the lower limit

Reply:
Got it, do you think that the existing slippage constraints are enough to protect against this? Client UX will typically display max slippage that the user could experience. If not, do you mind adding a comment in there with the above explanation about low number protection?

Reply:
This calculation is done for a deposit, where the slippage constraints go in the opposite direction, specifying the maximum to take, not the minimum, so unfortunately it doesn't cover this situation.  I'll add the comment with this info

Reply:
Comment add in in `4a936c6`